rust/src/transformations/manipulation/mod.rs

```diff
@@ -7,55 +7,96 @@ use opendp_derive::bootstrap;
 use crate::core::{Domain, Function, Metric, MetricSpace, StabilityMap, Transformation};
 use crate::domains::{AtomDomain, VectorDomain};
 use crate::error::*;
-use crate::metrics::{IntDistance, SymmetricDistance};
+use crate::metrics::{
+    ChangeOneDistance, HammingDistance, InsertDeleteDistance, IntDistance, SymmetricDistance,
+};
 use crate::traits::{CheckAtom, CheckNull, DistanceConstant};
 
+pub trait DatasetDomain: Domain {
+    type RowDomain: Domain;
+}
+
+impl<D: Domain> DatasetDomain for VectorDomain<D> {
+    type RowDomain = D;
+}
+
+pub trait DatasetMetric: Metric<Distance = IntDistance> {}
+
+impl DatasetMetric for SymmetricDistance {}
+impl DatasetMetric for InsertDeleteDistance {}
+impl DatasetMetric for ChangeOneDistance {}
+impl DatasetMetric for HammingDistance {}
+
+pub trait RowByRowDomain<DO: DatasetDomain>: DatasetDomain {
+    fn translate(&self, output_row_domain: DO::RowDomain) -> DO;
+    fn apply_rows(
+        value: &Self::Carrier,
+        row_function: &impl Fn(
+            &<Self::RowDomain as Domain>::Carrier,
+        ) -> Fallible<<DO::RowDomain as Domain>::Carrier>,
+    ) -> Fallible<DO::Carrier>;
+}
+
+impl<DIA: Domain, DOA: Domain> RowByRowDomain<VectorDomain<DOA>> for VectorDomain<DIA> {
+    fn translate(
+        &self,
+        output_row_domain: <VectorDomain<DOA> as DatasetDomain>::RowDomain,
+    ) -> VectorDomain<DOA> {
+        VectorDomain {
+            element_domain: output_row_domain,
+            size: self.size,
+        }
+    }
+
+    fn apply_rows(
+        value: &Self::Carrier,
+        row_function: &impl Fn(&DIA::Carrier) -> Fallible<DOA::Carrier>,
+    ) -> Fallible<Vec<DOA::Carrier>> {
+        value.iter().map(row_function).collect()
+    }
+}
+
 /// Constructs a [`Transformation`] representing an arbitrary row-by-row transformation.
-pub(crate) fn make_row_by_row<DIA, DOA, M>(
-    atom_input_domain: DIA,
-    atom_output_domain: DOA,
-    atom_function: impl 'static + Fn(&DIA::Carrier) -> DOA::Carrier,
-) -> Fallible<Transformation<VectorDomain<DIA>, VectorDomain<DOA>, M, M>>
+pub(crate) fn make_row_by_row<DI, DO, M>(
+    input_domain: DI,
+    input_metric: M,
+    output_row_domain: DO::RowDomain,
```

Comment:
The ordering of the arguments feels a little odd here. It seems like we usually have Domains together first. Is this to align with the new convention for partial construction?

Reply:
Yes, it is so that the signature support partials. ^^